### PR TITLE
PI: Check duplicate objects in writer._sweep_indirect_references

### PR DESCRIPTION
**Pull request by Hatell**
PdfFileWriter._sweepIndirectReferences checks if object is already in pdf.

This check will reduce file-size significantly when merge pdfs and there are same objects, for example images.

### CODE REVIEW

**Comment by claird** [comment]
Hello,

Thank you for your question, comments, and/or concerns regarding PyPDF2.
All of our PyPDF2 users are very important to us.  Unfortunately at this
present time we are in the process of regrouping our office.  Your
questions are very important to us but regretfully we will not be able to
address any issues regarding PyPDF2 until later this fall.  However, if
this is an urgent matter and you need immediate assistance please respond
to this email and we will try our best to accommodate your needs sooner.
My apologies for the inconvenience.  Thank you for your patience.

Sincerely,

Selma Kishwar

Phaseit, Inc.

selmaikishwar@gmail.com

**Comment by MartinThoma** [comment]
Hey! It's been 7 years, but your PR was not forgotten!

Would you mind to either re-do the PR or to fix the merge conflicts?

If not, I can take care of it and give credit to you via the Github co-authored-by feature

**Comment by Hatell** [comment]
I can take a look this soon.

**Comment by Hatell** [comment]
Now I think this is ready to testing.

I done one quick test and verified that merged PDF is smaller and looks same.

**Comment by Hatell** [comment]
Now tests should pass.

**Comment by MartinThoma** [comment]
I'm very sorry that I didn't merge it :see_no_evil: There were just too many open issues / PRs - i completely missed that one :see_no_evil:

**Comment by MartinThoma** [comment]
Would you mind doing the changes again with the latest `main` branch?

I would completely understand if you don't want to do that. In that case I would suggest that I take care of it and give the credit to you in the commit message. With Githubs "co-authored-by" feature you would also be seen as an author of the commit.

**Comment by Hatell** [comment]
I take a look.

**Comment by Hatell** [comment]
I think it should work now.

I did few tests and it worked well.

**Review by MartinThoma — APPROVED** [review]
Looks good to me

**Comment by MartinThoma** [comment]
@pubpub-zz @MasterOdin What do you think?

**Comment by MartinThoma** [comment]
@Hatell If nobody has doubts on that one, I'll merge + release latest on Sunday evening.

**Comment by MartinThoma** [comment]
I'm running a test over thousands of PDFs with essentially this code:

```python
from PyPDF2 import PdfReader, PdfWriter

reader = PdfReader(path)
writer = PdfWriter()

for page in reader.pages:
    writer.add_page(page)

writer.add_metadata(reader.metadata)
```

Then I check the file size before and after. When the difference is over 1 MB, I print it for manual inspection. Here is what I found so far:

**Comment by MartinThoma** [comment]
*   5,819,088 vs      966,396:    4,852,692 for pdf/04506b017e5ae60dd139b786c593006f.pdf: https://corpora.tika.apache.org/base/docs/govdocs1/917/917314.pdf
*  16,522,455 vs    3,587,895:   12,934,560 for pdf/08d27586646eee476631d55a3142300b.pdf: https://corpora.tika.apache.org/base/docs/govdocs1/924/924799.pdf
*  19,424,110 vs      473,147:   18,950,963 for pdf/0c4382438ea5b80073c8b71ddd89680b.pdf: https://corpora.tika.apache.org/base/docs/govdocs1/990/990065.pdf
*  14,681,925 vs      834,252:   13,847,673 for pdf/0f6d9402be03902014b6fbcdab3e558a.pdf: https://corpora.tika.apache.org/base/docs/govdocs1/922/922840.pdf
*  12,302,956 vs      630,459:   11,672,497 for pdf/0fb84cf340cd5a701610901057b04ba9.pdf: https://corpora.tika.apache.org/base/docs/govdocs1/982/982745.pdf
*   6,316,800 vs    1,011,880:    5,304,920 for pdf/1baa7c78e37697af70b51a2d3b0d9838.pdf: https://corpora.tika.apache.org/base/docs/govdocs1/921/921260.pdf
*   5,444,052 vs      158,680:    5,285,372 for pdf/25e639204d06f281cad6a70ab37a02fc.pdf: https://corpora.tika.apache.org/base/docs/govdocs1/938/938315.pdf
*  13,027,189 vs      821,442:   12,205,747 for pdf/2b091c791b64ba620cfd298c501ebf12.pdf: https://corpora.tika.apache.org/base/docs/govdocs1/922/922726.pdf
*  14,916,466 vs    1,661,672:   13,254,794 for pdf/2d0b14a363180f4f867e02eda8907c3f.pdf: https://corpora.tika.apache.org/base/docs/govdocs1/951/951777.pdf
*   6,109,775 vs      233,256:    5,876,519 for pdf/2e9a1bf7921413f4511cfb3e9ee1f23a.pdf: https://corpora.tika.apache.org/base/docs/govdocs1/914/914326.pdf
*   6,698,684 vs    1,883,640:    4,815,044 for pdf/30864c8b8e29386589481d1bc7582d50.pdf: https://corpora.tika.apache.org/base/docs/govdocs1/949/949428.pdf
*   8,804,687 vs    4,278,708:    4,525,979 for pdf/35c59e0be909ee88508262ba664fc6b7.pdf: https://corpora.tika.apache.org/base/docs/govdocs1/911/911140.pdf
*   6,032,308 vs    1,530,944:    4,501,364 for pdf/3d31542c851b1a8e7db920ba90541cb3.pdf: https://corpora.tika.apache.org/base/docs/govdocs1/921/921496.pdf

**Comment by MartinThoma** [comment]
@Hatell Using the approach above, a few files failed that were working before:

* https://corpora.tika.apache.org/base/docs/govdocs1/906/906769.pdf: 'NoneType' object has no attribute 'hash_value'
* https://corpora.tika.apache.org/base/docs/govdocs1/946/946956.pdf: 'NoneType' object has no attribute 'hash_value'

**Comment by Hatell** [comment]
Thanks for tests. I check what happens in those files.

File: PyPDF2/_writer.py [hunk 1]
```diff
@@ -868,8 +869,13 @@ def _sweep_indirect_references(
                 if newobj is None:
                     try:
                         newobj = data.pdf.get_object(data)
+                        hash_value = newobj.hash_value()
+                        # Check if object is already added to pdf.
+                        if hash_value in self._idnum_hash:
+                            return IndirectObject(self._idnum_hash[hash_value], 0, self)
                         self._objects.append(None)  # placeholder
                         idnum = len(self._objects)
+                        self._idnum_hash[hash_value] = idnum
```
Comment (MasterOdin) [inline PyPDF2/_writer.py:878]: Should we be doing a similar operation [here](https://github.com/py-pdf/PyPDF2/blob/fe7b4c9df5bd65be7eeeee252ebea4862389f3e9/PyPDF2/_writer.py#L734) where the `extern_map` object is original setup?

**Review by MasterOdin — COMMENTED** [review]
It feels like the usage of `self._idnum_hash` and `extern_map` share a slightly similar purpose, that could we get rid of the former altogether? With this change, could it be that cases where we would avoid the `if` surrounding `if` statement of the change we'll now enter as we've not written to `extern_map` for those deep self-referential objects.

**Comment by Hatell** [comment]
Now those NoneType object has no attribute is handled.

File: PyPDF2/_writer.py (hunk 1 above)
Comment (Hatell) [reply PyPDF2/_writer.py]: I'm not sure.

Comment (MartinThoma) [reply PyPDF2/_writer.py]: @MasterOdin I think this PR already adds quite a bit of value and Hatell already rebased twice due to other changes. In order to Hatells side low, I tend to merge soon without the de-duplicatoin with extern_map. We can still do it later (depending on how complex it is, I might do it directly after merging). Would that be ok for you?

Comment (MasterOdin) [reply PyPDF2/_writer.py]: That's fine.

**Pull request merged**